stix/idl/processing/spectrogram/stx_science_data_lightcurve.pro

```diff
@@ -83,110 +90,163 @@ function stx_science_data_lightcurve, fits_path, energy_ranges = edges_in,  time
+  data_all = []
+  error_all = []
+  ltime_all = []
+  ut2_time_all = []
+  duration_all = []
+  rcr_all = []
+  for this_file = 0,nfiles-1 do begin
+
+    if strpos(orig_filename, 'cpd') gt -1 or strpos(orig_filename, 'xray-l1') gt -1 then begin
+      stx_convert_pixel_data, fits_path_data = fits_path[this_file], fits_path_bk =  fits_path_bk, distance = distance, time_shift = time_shift, ospex_obj = ospex_obj, $
+        det_ind = det_ind, pix_ind = pix_ind, sys_uncert = sys_uncert, plot = 0, _extra= _extra
+    endif else if strpos(orig_filename, 'spec') gt -1 or strpos(orig_filename, 'spectrogram') gt -1 then begin
+      stx_convert_spectrogram, fits_path_data = fits_path[this_file], fits_path_bk =  fits_path_bk, distance = distance, time_shift = time_shift, ospex_obj = ospex_obj, $
+        sys_uncert = sys_uncert, plot = 0, _extra= _extra
+      if keyword_set(det_ind) or keyword_set(pix_ind) then  message, 'ERROR: Detector and pixel selection not possible with spectrogram files.'
+    endif else begin
+      message, 'ERROR: the FILENAME field in the primary header should contain either cpd, xray-l1 or spec'
+    endelse
   
-  ;use OSPEX bin_data method to bin counts in given energy bands
-  energy_summed_counts = data_obj->bin_data(data = counts_str, intervals = energy_ranges, $
-    eresult = energy_summed_error, ltime = energy_summed_ltime)
-
-  energy_summed_str = {data:energy_summed_counts, edata:energy_summed_error, ltime:energy_summed_ltime}
-
-  ; determine time bins with minimum duration - keep adding consecutive bins until the minimum
-  ; value is at least reached
-  i=0
-  j=0
-  total=0
-  iall=[]
-
-  while (i lt n_elements(duration)-1) do begin
-    while (total lt time_min)  and (i+j le n_elements(duration)-1) do begin
-      total = total(duration[i:i+j])
-      j++
+    data_obj = ospex_obj->get(/obj,class='spex_data')
+    counts_str = ospex_obj->getdata(spex_units='counts')
+  
+    ut_time = ospex_obj->getaxis(/ut, /edges_1)
+    ut2_time = ospex_obj->getaxis(/ut, /edges_2)
+    ;ut2_time_all = ut2_time
+    duration = ospex_obj->getaxis(/ut, /width)
+    filters_all = ospex_obj->get(/spex_interval_filter) 
+    
+    ;use OSPEX bin_data method to bin counts in given energy bands
+    energy_summed_counts = data_obj->bin_data(data = counts_str, intervals = energy_ranges, $
+      eresult = energy_summed_error, ltime = energy_summed_ltime)
+  
+    energy_summed_str = {data:energy_summed_counts, edata:energy_summed_error, ltime:energy_summed_ltime}
+  
+    ; determine time bins with minimum duration - keep adding consecutive bins until the minimum
+    ; value is at least reached
+    i=0
+    j=0
+    total=0
+    iall=[]
+  
+    while (i lt n_elements(duration)-1) do begin
+      while (total lt time_min)  and (i+j le n_elements(duration)-1) do begin
+        total = total(duration[i:i+j])
+        j++
+      endwhile
+      iall = [iall,i]
+      i = i+j
+      j = 0
+      total = 0
     endwhile
```

Comment:
If speed becomes an issue where I think this is essentially a histogram operation with a bin width of time_min and use the reverse indices e.g. `h = histogram(duration, binsize = time_min, reverse=r, locations=bins)` which would be significantly faster.

Reply:
Thanks for the suggestion. This is actually a piece of code from @grazwegian, not sure why here it is marked in green since I've not touched it. Maybe because it is now in the for loop iterating all files. Anyway, what is making the code slower is the OSPEX reading of the FITS files (see lines 113-117), so not sure we can do something here in this code for this

Reply:
Thanks, I will take a look into that improvement. I expect that the OSPEX reading could also be made a lot faster as there are a lot of things we don't need if we just want to generate the lightcurve without doing any spectral analysis.